### ISSUE TITLE
Upgrading rpm causes /usr/local/bin/leofs-adm to disappear on EL6/7

### ISSUE THREAD

**Issue by vstax** [leo-project/leofs#742]
When upgrading leofs rpm directly (rpm -U) instead of recommended way of installing new one in parallel and removing old one later - which can be handy during development - leofs-adm alternative is removed, causing /usr/local/bin/leofs-adm to disappear.

rpm package needs some script change, it seems. Debian package is not affected since it doesn't use alternatives (and doesn't support installing in parallel). I'll look into cause and fix this a bit later.

**Comment by mocchira** [comment]
@vstax confirmed. thanks!

> rpm package needs some script change, it seems. Debian package is not affected since it doesn't use alternatives (and doesn't support installing in parallel). I'll look into cause and fix this a bit later.

Can I assign this issue to you?
(You wouldn't need to be worried about the release time frame.)

**Comment by vstax** [comment]
@mocchira Sure.

This problem is specific to same-version upgrades with only revision change, e.g. leofs-1.3.4-1 -> leofs-1.3.4-2. Original use of alternatives there assumed that version always changes, e.g. old version is in `/usr/local/leofs/1.3.3` and new in `/usr/local/leofs/1.3.4`. It fails when it's upgrade with both directories (old and new version) being the same. Users shouldn't be affected by this bug, at least not for now (unless new release of packages for the same leofs version is released).

To fix this, `update-alternatives --remove leofs-adm` command shouldn't be executed in case of upgrade with above conditions (but executed in all other cases). I don't know a way of handling this cleanly, as rpm doesn't pass all the required information to scripts to differentiate this case. Some hackish way to check if version being upgraded goes into the same directory as the new one is needed, I'm still thinking on how to do it the best way for now.

**Comment by mocchira** [comment]
Thanks for in detail.
I'd drop a comment if I find some clean way to handle this particular case.

**Comment by vstax** [comment]
I've created https://github.com/leo-project/leofs_package/pull/5

Not really like idea of using %posttrans, but, well, it seems to be clean enough. It works by making "update-alternatives --install" execute at the very end so it's always executed after "update-alternatives --remove" of previous package. So during same version upgrade the same alternative is removed and then installed again, but it solves the problem.

Tested:
1.3.3 built with older package version + 1.3.4 built with new package (with this change) - parallel installation of new version + erase of older version; also upgrade
1.3.3 built with new package + 1.3.4 built with new package - parallel installation of new version + erase of older version; also upgrade
1.3.4-1 built with older package upgraded to 1.3.4-2 built with new package
1.3.4-1 built with new package upgraded to 1.3.4-2 built with new package
all work as expected. Downgrades work fine as well (except for 1.3.4-2 to 1.3.4-1 when 1.3.4-1 is built from older package - this one isn't supposed to work).

If the 1.3.4-1 -> 1.3.4-2 upgrade was performed before this change and alternative doesn't exist currently there will be non-fatal error during uninstallation of older package, but I don't really think anyone will experience this, except for on purpose.

So, I don't think there is any need to document anything, this problem (if fixed now) shouldn't affect anyone.
